rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,201 @@
+  const state = useSample(props, ref);
+  useSampleStyles(state);
+
+  return renderSample(state);
+});
+```
+
+#### useSample.ts
+
+Accepts the component props and handles and internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps) => {
+  const resolvedRef = ref || React.useRef();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+```typescript
+import { getSlots } from '@fluentui/react-utils'
+
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.otherSlot {...slotProps.otherSlot} />
+    </slots.root>
+  );
+};
+```
+
+The `state` stores all information on the slots/tags/JSX that should be rendered after processing by the relevant hook. The `getSlots` utility extracts known slot information such as tag names and props.
+
+See below for how `state` might contain useful props for rendering.
+
+#### Sample.types.ts
+The below will probably be present in every component, other utility types are fair game if they are necessary.
+
+```typescript
+import * as React from 'react';
```

Comment:
**q (out off topic):** Are we gonna go with this obsolete pattern ? (if we use this just for types, there might be issues with tree shaking - in convergence, I'd suggest to use `esModuleInterop: true` for TS config -> babel will exec transpilation ). 

```
// common standard
import React from 'react'
// explicit type imports - zero runtime
import type {...} from 'react'

```

Reply:
https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

> Change all default React imports (i.e. import React from "react") to destructured named imports (ex. import { useState } from "react") which is the preferred style going into the future. This codemod will not affect the existing namespace imports (i.e. import * as React from "react") which is also a valid style. The default imports will keep working in React 17, but in the longer term we encourage moving away from them.

React team performs small steps to go with ESM. I think that we should rather avoid `default` imports

Reply:
agree with avoiding default import (although core team provides codemod that will transform default to explicit). I'd suggest to be rather be explicit with all API's (not just react) -> thus using named exports

Reply:
I don't think this is too relevant to this RFC, objective is to make the workflow clearer while there is room to make slight sanity improvements this goes a bit deeper IMO

Reply:
yup, I explicitly stated it's out of topic, just wanted to get additional info for broader audience :) thx!